core/src/main/scala/org/apache/spark/scheduler/BlacklistTracker.scala

```diff
@@ -0,0 +1,165 @@
+import org.apache.spark.Success
+import org.apache.spark.TaskEndReason
+import org.apache.spark.annotation.DeveloperApi
+import org.apache.spark.util.SystemClock
+import org.apache.spark.util.ThreadUtils
+import org.apache.spark.util.Utils
+
+
+/**
+ * BlacklistTracker is design to track problematic executors and node on application level.
+ * It is shared by all TaskSet, so that once a new TaskSet coming, it could be benefit from
+ * previous experience of other TaskSet.
+ *
+ * Once task finished, the callback method in TaskSetManager should update failedExecutorMap.
+ */
+class BlacklistTracker(sparkConf: SparkConf) {
+  // maintain a ExecutorId --> FailureStatus HashMap
+  private val failedExecutorMap: mutable.HashMap[String, FailureStatus] = mutable.HashMap()
+
+  // Apply Strategy pattern here to change different blacklist detection logic
+  private val strategy = BlacklistStrategy(sparkConf)
+
+  // A daemon thread to expire blacklist executor periodically
+  private val scheduler = ThreadUtils.newDaemonSingleThreadScheduledExecutor(
+      "spark-scheduler-blacklist-expire-timer")
+
+  private val clock = new SystemClock()
+
+  def start(): Unit = {
+    val scheduleTask = new Runnable() {
+      override def run(): Unit = {
+        Utils.logUncaughtExceptions(expireExecutorsInBlackList(failedExecutorMap))
+      }
+    }
+    scheduler.scheduleAtFixedRate(scheduleTask, 0L, 60, TimeUnit.SECONDS)
+  }
+
+  def stop(): Unit = {
+    scheduler.shutdown()
+    scheduler.awaitTermination(10, TimeUnit.SECONDS)
+  }
+
+  // The actual implementation is delegated to strategy
+  private def expireExecutorsInBlackList(
+      failureExecutors: mutable.HashMap[String, FailureStatus]): Unit = synchronized {
+    strategy.expireExecutorsInBlackList(failureExecutors)
+  }
+
+  def updateFailureExecutors(info: TaskInfo, reason: TaskEndReason) : Unit = synchronized {
+    reason match {
+      // If task succeeding, remove related record from failedExecutorMap
+      case Success =>
+        removeFailureExecutors(info.executorId, Some(info.taskId))
+
+      // If task failing, update latest failure time and failedTaskIds
+      case _ =>
+        val executorId = info.executorId
+        failedExecutorMap.get(executorId) match {
+          case Some(failureStatus) =>
+            failureStatus.updatedTime = clock.getTimeMillis()
+            val failedTimes = failureStatus.failedTasks.getOrElse(info.taskId, 0) + 1
+            failureStatus.failedTasks.update(info.taskId, failedTimes)
+          case None =>
+            val failedTasks = mutable.HashMap(info.taskId -> 1)
+            val failureStatus = new FailureStatus(
+              clock.getTimeMillis(),
+              info.host,
+              failedTasks)
+            failedExecutorMap.update(executorId, failureStatus)
+        }
+    }
+  }
+
+  def removeFailureExecutors(
+      executorId: String,
+      taskId: Option[Long]) : Unit = synchronized {
+        taskId match {
```

Comment:
similar comment here on `taskId` -- I think this is more naturally just two different methods, one with a `taskId`, and one without, since the callsite always knows if it has a taskId or not.